R/disease_progression.R

```diff
@@ -84,3 +84,49 @@ update_to_asymptomatic_infection <- function(
     )
   }
 }
+
+#' @title Sub-patent progression for vivax model
+#' @description
+#' randomly selects individuals to clear sub-patent infections based on impact of 
+#' individual immunity to detectability
+#'
+#' @param state the human state variable
+#' @param variables the available human variables
+#' @param parameters model parameters
+#' @noRd
+create_subpatent_progression_process <- function(
+  state,
+  variables,
+  parameters
+  ){
+  function(timestep) {
+    ## Get rates for each subpatent individual
+    subpatent_index <- state$get_index_of("U")
+    rates <- subpatent_duration(parameters, variables, index = subpatent_index)
+    ## Get individuals who are going to change
+    to_move <- subpatent_index$sample(1/rates)
+    # tupdate individuals and infectivity
+    update_infection(
+      state,
+      to_state = "S",
+      infectivity = variables$infectivity,
+      new_infectivity = 0,
+      to_move
+    )
+  }
+}
+
+#' @title Sub-patent duration for vivax model
+#' @description
+#' calculates sub-patent duration based on detectability (anti-parasite) immunity
+#'
+#' @param parameters model parameters
+#' @param variable immunity variables
+#' @param index index of subpatent infected individuals 
+#' @noRd
+subpatent_duration <- function(parameters, variables, index){
```

Comment:
Looking at White (2018) supplementary figure 1.
I think this corresponds to Ap (anti-parasite immunity)?
If so I think we should call it that so it references to the paper. I also think you'll call this function in more than one place (the Ap circle is on two flows in that diagram), so good to generalise.

It would actually be good to check if the differently named immunity functions in the Pf and Pv exactly align (for example is vivax anti-parasite immunity == falciparum blood_immunity?). If they do then I think we would want to make sure we don't have identical code with differnt names - ie we would want to use a single function for both Pf and Pv

Reply:
The section does correspond with anti-parasite immunity, but there are two separate things that are being modelled: duration of sub-patent infection (rate of recovery to S) and probability of patent infection phi_lm. It's on two places in the equation but the modelling is slightly different in each one: the equations mirror each other but have different input parameters; so I'm happy to generalise that. In this pull request I've only the duration, not the probability of patent infection (I actually thought that I already done this, but I think I lost that code trying to discretise the work into smaller chunks! I'll add that back in with the next push).

![Screenshot from 2023-06-29 12-12-50](https://github.com/mrc-ide/malariasimulation/assets/73181123/7b8c47e6-3671-4873-9ca8-3adac053117d)

The falciparum and vivax don't align exactly (also, I think we said that the anti-parasite immunity maps onto the detectability immunity (ID), rather than the blood immunity (IB)): the falciparum has what looks like a hill function within another hill function, and only uses acquired ID, not both acquired and maternal ID. It's also conceptually different, because it's the probability that an asymptomatic infection is detectable, so doesn't get used in the same way.

In terms of generalisation, though, there are a number of discrete functions that have the same structure (basically all the hill functions): blood_immunity and clinical_immunity, then probability of detection and severe immunity are both double hill functions. It could be worth generalising those at some point...?

So: I'm going to make a general AP_immunity function that can take parameters from either sub-patent duration or probability of patent infection. Then this can be used in both places.